File: 해시/베스트앨범/베스트앨범.js
function solution(genres, plays) {
    const objByGenre = {}
    const score = {}

    genres.forEach((genre, i) => {
        if (objByGenre[genre]) {
            objByGenre[genre].push([i, plays[i]])
            score[genre] += plays[i]
        } else {
            objByGenre[genre] = [[i, plays[i]]]
            score[genre] = plays[i]
        }
    })

    const result = []
    Object.entries(score)
        .sort((a, b) => b[1] - a[1])
        .forEach((v) => {
            const [genre, counts] = v
            result.push(
                ...objByGenre[genre]
                    .sort((a, b) => b[1] - a[1])
                    .slice(0, 2)
                    .map((v) => v[0])
            )
        })
    return result
}

File: 프로그래머스/3/42579. 베스트앨범/베스트앨범.java
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class Solution {
	public int[] solution(String[] genres, int[] plays) {
		HashMap<String, Integer> score = new HashMap<>();
		HashMap<String, ArrayList<int[]>> genresMap = new HashMap<>();

		for (int i = 0; i < genres.length; i++) {
			String genre = genres[i];
			int play = plays[i];

			if (genresMap.containsKey(genre)) {
				genresMap.get(genre).add(new int[] {i, play});
				score.put(genre, score.get(genre) + play);
			} else {
				ArrayList<int[]> list = new ArrayList<>();
				list.add(new int[] {i, play});
				genresMap.put(genre, list);
				score.put(genre, play);
			}
		}

		ArrayList<Integer> result = new ArrayList<>();
		List<Map.Entry<String, Integer>> sortedScore = new ArrayList<>(score.entrySet());
		sortedScore.sort((a, b) -> b.getValue().compareTo(a.getValue()));

		for (Map.Entry<String, Integer> entry : sortedScore) {
			ArrayList<int[]> list = genresMap.get(entry.getKey());
			list.sort((a, b) -> b[1] - a[1]);
			for (int i = 0; i < Math.min(2, list.size()); i++) {
				result.add(list.get(i)[0]);
			}
		}

		return result.stream().mapToInt(i -> i).toArray();
	}
}
